### PR TITLE
handle alerts in audio cue service, add `List Alerts` command and settings

### LINKED ISSUE #195282: Consider `alert` as option instead of audio cue for components
@jooyoungseo and @rperez030 brought to my attention the fact that audio cues might not be practical for certain users. 

We should consider adding `alert` when audio cues are disabled for various features.

### LINKED ISSUE #201910: Add `List Alerts` for overview of alerts and their status
The command `List Audio Cues` allows for reviewing the available cues, configuring each, and checking the status (enabled/disabled). Now that we will have many configurable alerts, we should have a similar command for these. 

I will mention these commands in the help dialog, which I believe solves the issue of users not knowing how to disable them. 

cc @rperez030 and @jooyoungseo

### PR DESCRIPTION
**Pull request by meganrogge**
fixes #195282
fixes #201910

### CODE REVIEW

**Comment by hediet** [comment]
Thanks for integrating the AccessibleNotificationService into the AudioCueService!

The PR is quite hard to review though, because it is a refactoring of the AccessibleNotificationService.
Can you try to revert your initial PR where you introduced AccessibleNotificationService and then rebase this PR onto that?
That would make reviewing so much easier and is what I initially suggested. Thank you!

**Review by hediet — CHANGES_REQUESTED** [review]
Please revert the introduction of IAccessibleNotificationService in a separate PR and rebase this PR again, so that it only contains the changes for the feature.

Now it is very hard to review the changes needed for the feature.

**Comment by meganrogge** [comment]
@hediet that would require reverting all 4 of these PRS, which would complicate things quite a bit. 

https://github.com/microsoft/vscode/pulls?q=is%3Apr+author%3Ameganrogge+is%3Aclosed+accessible+notification+service

I'm happy to walk through the changes with you if you want on a call. Essentially, now I call the audio cue service instead of the accessible notification service.

File: src/vs/platform/audioCues/browser/audioCueService.ts [hunk 1]
```diff
@@ -43,13 +45,17 @@ export class AudioCueService extends Disposable implements IAudioCueService {
 	constructor(
 		@IConfigurationService private readonly configurationService: IConfigurationService,
 		@IAccessibilityService private readonly accessibilityService: IAccessibilityService,
-		@ITelemetryService private readonly telemetryService: ITelemetryService,
-	) {
+		@ITelemetryService private readonly telemetryService: ITelemetryService) {
```
Comment (hediet) [inline src/vs/platform/audioCues/browser/audioCueService.ts:48]: Nit (not important to fix): the trailing comma and uniform formatting in the original code before this change makes diffs and editing (using multi-cursor) easier :)

File: src/vs/platform/audioCues/browser/audioCueService.ts [hunk 2]
```diff
@@ -306,21 +369,29 @@ export class AudioCue {
 		name: localize('audioCues.lineHasError.name', 'Error on Line'),
 		sound: Sound.error,
 		settingsKey: 'audioCues.lineHasError',
+		alertSettingsKey: AccessibilityAlertSettingId.Error,
```
Comment (hediet) [inline src/vs/platform/audioCues/browser/audioCueService.ts:372]: Can we combine the alertSettingsKey and the settingsKey in some future refactoring?
Maybe sth like this:
```json
"signals.lineHasError": {
   "audioCue": "on",
   "alert": "on",
}
```

File: src/vs/platform/audioCues/browser/audioCueService.ts [hunk 3]
```diff
@@ -152,15 +165,15 @@ export class AudioCueService extends Disposable implements IAudioCueService {
 		Event.filter(this.configurationService.onDidChangeConfiguration, (e) =>
 			e.affectsConfiguration('audioCues.enabled')
 		),
-		() => /** @description config: audioCues.enabled */ this.configurationService.getValue<'on' | 'off' | 'auto'>('audioCues.enabled')
+		() => /** @description config: audioCues.enabled */ this.configurationService.getValue<'on' | 'off' | 'auto' | 'userGesture' | 'always' | 'never'>('audioCues.enabled')
 	);
 
-	private readonly isEnabledCache = new Cache((cue: AudioCue) => {
+	private readonly isCueEnabledCache = new Cache((cue: AudioCue, userGesture?: boolean) => {
```
Comment (hediet) [inline src/vs/platform/audioCues/browser/audioCueService.ts:171]: Unfortunately, the `Cache` doesn't work like this - you can only set one parameter.
I think multi parameter cache makes sense, so feel free to add this functionality to the Cache.

I just wonder why the compiler didn't complain here.

File: src/vs/platform/audioCues/browser/audioCueService.ts [hunk 4]
```diff
@@ -184,12 +199,35 @@ export class AudioCueService extends Disposable implements IAudioCueService {
...
+				!this.screenReaderAttached.read(reader)
+			) {
+				return false;
+			}
+			return setting === true || setting === 'always' || setting === 'userGesture' && userGesture;
+		});
+	});
+
+	public isAlertEnabled(cue: AudioCue): boolean {
+		return this.isAlertEnabledCache.get(cue).get() ?? false;
```
Comment (hediet) [inline src/vs/platform/audioCues/browser/audioCueService.ts:222]: Fishy: Enablement does not depend on `userGesture`. See comment above.

File: src/vs/platform/audioCues/browser/audioCueService.ts [hunk 5]
```diff
@@ -29,6 +30,7 @@ export interface IAudioCueService {
 export interface IAudioCueOptions {
 	allowManyInParallel?: boolean;
 	source?: string;
+	userGesture?: boolean;
```
Comment (hediet) [inline src/vs/platform/audioCues/browser/audioCueService.ts:33]: Can you document what `useGesture` means and what problem it solves?

**Review by hediet — CHANGES_REQUESTED** [review]
See comments

**Comment by hediet** [comment]
> I initially created the AccessibleNotificationService to handle audio cues and alerts in one place and to deal with the complexity introduced by the format/save cues which have setting values of userGesture, always, or never.

I understand. I do think however that the complexity the `AccessibleNotificationService` added (having a separate system that is very similar to AudioCues, with parallel settings and events) is higher than the complexity it removed (dealing with `userGesture` etc). Also, `userGesture` could be relevant for audio cues as well.

Btw. thank you for all your work on audio cues! :) I really appreciate that you push this feature.

**Comment by meganrogge** [comment]
> I understand. I do think however that the complexity the AccessibleNotificationService added (having a separate system that is very similar to AudioCues, with parallel settings and events) is higher than the complexity it removed (dealing with userGesture etc). Also, userGesture could be relevant for audio cues as well.

I agree, which is why I made the change that you suggested :)

> If you like we can schedule a meeting these days to go through the code and discuss how the AudioCue service could evolve to account for the new and future requirements.

Sounds good. Do you think we should do this now or as new and future requirements arise?

File: src/vs/platform/audioCues/browser/audioCueService.ts (hunk 2 above)
Comment (meganrogge) [reply src/vs/platform/audioCues/browser/audioCueService.ts:372]: https://github.com/microsoft/vscode/issues/202018

**Comment by hediet** [comment]
> Sounds good. Do you think we should do this now or as new and future requirements arise?

I'm fine with either, as you have motivation and time!

**Comment by meganrogge** [comment]
I have other stuff going on this iteration, so maybe in February we should meet.

**Comment by meganrogge** [comment]
I think the `userGesture` is still not being respected 🤔 . I find the cache pretty confusing

File: src/vs/platform/audioCues/browser/audioCueService.ts [hunk 6]
```diff
@@ -254,17 +260,17 @@ function playAudio(url: string, volume: number): Promise<HTMLAudioElement> {
...
-	constructor(private readonly getValue: (value: TArg) => TValue) {
+	private readonly map = new Map<{ arg: TArg; optionalArg: any }, TValue>();
+	constructor(private readonly getValue: (value: TArg, optionalArg?: any) => TValue) {
 	}
 
-	public get(arg: TArg): TValue {
-		if (this.map.has(arg)) {
-			return this.map.get(arg)!;
+	public get(arg: TArg, optionalArg?: any): TValue {
+		if (this.map.has({ arg, optionalArg })) {
```
Comment (hediet) [inline src/vs/platform/audioCues/browser/audioCueService.ts:268]: This creates a new object every time. Because `Map` uses referential equality, this will always be false.
This is a little bit tricky.
I suggest to remove `optionalArg`, instantiate `TArg` with `{ cue: AudioCue, userGesture: boolean }` at the use-side and add an optional `getKey` function to the `Cache` class which would be used for adding the value to the map.
The `getKey` function could be `JSON.stringify([cue.id, userGesture])`.

File: src/vs/platform/audioCues/browser/audioCueService.ts [hunk 7]
```diff
@@ -260,17 +264,22 @@ function playAudio(url: string, volume: number): Promise<HTMLAudioElement> {
 }
 
 class Cache<TArg, TValue> {
-	private readonly map = new Map<{ arg: TArg; optionalArg: any }, TValue>();
-	constructor(private readonly getValue: (value: TArg, optionalArg?: any) => TValue) {
+	private readonly map = new Map<TArg, TValue>();
+	constructor(private readonly getValue: (value: TArg) => TValue) {
+	}
+
+	private _getKey(event: IAudioCueEvent): string {
```
Comment (hediet) [inline src/vs/platform/audioCues/browser/audioCueService.ts:271]: Please don't refer to `IAudioCueEvent` from this general purpose utility.
It makes this code more complicated to understand (and fragile), as it is no longer an isolated unit, but intertangled with the non-general busisness logic. It also makes the business logic more complicated, as `Cache` is no longer a predictable component, but surprisingly special cased towards the business logic.

Instead, inject this function through the constructor (`private readonly getKey: (value: TArg) => unknown`).
That moves the key function to the right place and detangles the business logic from the general purpose part.

Generally, `as` casts are highly suspicious and can be usually avoided by a better design.

Whenever business logic gets more involved, I can only recommend to extract isolatable patterns as self-explanatory utilities. This makes the business logic more compact, as long as the function of the utility can be derived from its name ([a very good test here is to generate the class by copilot/chatgpt, when you only give it the API](https://chat.openai.com/share/63f7e45e-89f1-42ef-a5e3-069d7d49df53)).

File: src/vs/platform/audioCues/browser/audioCueService.ts [hunk 8]
```diff
@@ -220,22 +220,26 @@ export class AudioCueService extends Disposable implements IAudioCueService {
...
-		return this.isCueEnabledCache.get(cue, userGesture).get() ?? false;
+		return this.isCueEnabledCache.get({ cue, userGesture }).get() ?? false;
 	}
 
 	public onEnabledChanged(cue: AudioCue): Event<void> {
-		return Event.fromObservableLight(this.isCueEnabledCache.get(cue));
+		return Event.fromObservableLight(this.isCueEnabledCache.get({ cue }));
 	}
 }
+export interface IAudioCueEvent {
```
Comment (hediet) [inline src/vs/platform/audioCues/browser/audioCueService.ts:239]: This type can be inlined once the problem below is addressed.

**Comment by meganrogge** [comment]
still wip

File: src/vs/platform/audioCues/browser/audioCueService.ts [hunk 9]
```diff
@@ -268,8 +268,8 @@ class Cache<TArg, TValue> {
 	constructor(private readonly getValue: (value: TArg) => TValue) {
 	}
 
-	private _getKey(event: IAudioCueEvent): string {
-		return `${event.cue.name}-${event.userGesture}`;
+	private _getKey(arg: TArg): string {
+		return JSON.stringify(arg);
```
[the remark this answers is not in the document]
Comment (hediet) [reply src/vs/platform/audioCues/browser/audioCueService.ts:269]: Also, `TArg` in the map would just be `unknown` then.

[the remark this answers is not in the document]
Comment (meganrogge) [reply src/vs/platform/audioCues/browser/audioCueService.ts]: yes, just did this. but not sure how to avoid this cast https://github.com/microsoft/vscode/blob/93b603aada88cbf60731adf296ab114736695085/src/vs/platform/audioCues/browser/audioCueService.ts#L279

[the remark this answers is not in the document]
Comment (meganrogge) [reply src/vs/platform/audioCues/browser/audioCueService.ts]: figured it out

File: src/vs/platform/audioCues/browser/audioCueService.ts [hunk 10]
```diff
@@ -236,10 +236,7 @@ export class AudioCueService extends Disposable implements IAudioCueService {
 		return Event.fromObservableLight(this.isCueEnabledCache.get({ cue }));
 	}
 }
-export interface IAudioCueEvent {
-	readonly cue: AudioCue;
-	readonly userGesture?: boolean;
-}
+export interface IAudioCueEvent { readonly cue: AudioCue; readonly userGesture?: boolean }
```
Comment (hediet) [inline src/vs/platform/audioCues/browser/audioCueService.ts:244]: I'm sorry, but with inlining I meant to replace all occurrences of `IAudioCueEvent` with the its definition (`{ readonly cue: AudioCue; readonly userGesture?: boolean }`).
We can do this now, because the is only one occurrence of it. Also, because after inlining this type the type is very local and only passed into the cache, we can omit `readonly` to make it even shorter.

File: src/vs/platform/audioCues/browser/audioCueService.ts [hunk 11]
```diff
@@ -205,7 +205,7 @@ export class AudioCueService extends Disposable implements IAudioCueService {
 		});
 	}, (event) => this._getKey(event));
```
Comment (hediet) [inline src/vs/platform/audioCues/browser/audioCueService.ts:206]: Even better:
```
}, event => JSON.stringify(event));
```
You can even shorten this to:
```
}, JSON.stringify);
```

File: src/vs/platform/audioCues/browser/audioCueService.ts [hunk 12]
```diff
@@ -174,7 +174,7 @@ export class AudioCueService extends Disposable implements IAudioCueService {
 		() => /** @description config: audioCues.enabled */ this.configurationService.getValue<'on' | 'off' | 'auto' | 'userGesture' | 'always' | 'never'>('audioCues.enabled')
 	);
 
-	private readonly isCueEnabledCache = new Cache((event: IAudioCueEvent) => {
+	private readonly isCueEnabledCache = new Cache((event: { readonly cue: AudioCue; readonly userGesture?: boolean }) => {
```
Comment (hediet) [inline src/vs/platform/audioCues/browser/audioCueService.ts:177]: Now, you could shorten the type to
```ts
{ cue: AudioCue; userGesture?: boolean }
```
or even use a different type (and adjust callers to `getValue`):
```ts
[AudioCue, userGesture?: boolean]
```
... but that is just icing on the cake.

**Review by hediet — APPROVED** [review]
Thank you, that looks much better now!

By just looking at the code, moving out `JSON.stringify` from `Cache` to the use-sides looks like a complication of the code, but it actually reduces complexity, as you need less assumptions to understand `Cache`.

File: src/vs/platform/audioCues/browser/audioCueService.ts (hunk 12 above)
Comment (meganrogge) [reply src/vs/platform/audioCues/browser/audioCueService.ts:177]: I tried this to no avail 🤔

**Comment by meganrogge** [comment]
Thanks for all of the reviews 👍🏼
